databroker/broker.py

```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+known_special_keys = {
+    # The duration of the scan
+    'duration': lambda header: (datetime.fromtimestamp(header.stop.time) -
+                                datetime.fromtimestamp(header.start.time)),
+    # The number of events per event descriptor
+    'num_events': lambda header: [
+        len(list(mdsc.get_events_generator(descriptor, False)))
+        for descriptor in header.descriptors],
+}
+
+def _get_uid(key, document):
+    try:
+        # see if we have a hyphen
+        key, num = key.split('-')
+        num = int(num)
+    except ValueError:
+        # Nope! Must be asking for the whole uid
+        num = None
+    return document.uid[:num]
+
+
+@singledispatch
```

Comment:
I do not see what the single dispatch is helping with here.  In `summarize` it looks like you should only ever see `Headers` and in the `Header` `_get_value` it only every sees `Documents`.